### PR TITLE
Improve single document mode to address classic notebook usage cases

### PR DESCRIPTION
**Pull request by ellisonbg**
This is a beginning to improve JupyterLab's single document mode to address the usage cases of the classic notebook. In particular, this PR simplifies the single document layout and adds a document title header to the page. The user can switch from the normal JupyterLab layout (which we call multiple document mode) to this single document mode using Command+Shift+D or the View menu. The result is a user experience that is closer to the intent of the classic notebook by being more "document oriented." Here is an animated GIF of the transition:

jlab-sdm

## References

https://github.com/jupyterlab/jupyterlab/issues/8292

## Code changes

Adds a new title panel to the overall layout of the shell, between the header and top panels. Currently, this is a private shell panel, with no public extension points.

## Backwards-incompatible changes

None.

### CODE REVIEW

**Comment by ellisonbg** [comment]
I should note, there are a ton of other things we will want to do to improve single document mode, this PR is only meant to get us going. I will open an issue to track the bigger picture.

@afshin @vidartf @blink1073 and others who are interested in this work.

**Comment by telamonian** [comment]
I dig your new header styling. Should seem very welcoming to existing nb classic users.

Suggestion: single-document mode should keep its own separate "memory" of the sidebar state, which should default to no sidebars widgets showing. That way

- "single-document mode" is a more honest descriptor; it would at least start out as "just the single document, nothing else"
- greater similarity in layout with nb classic, while still retaining muted touches of jlab (eg the sidebars themselves will still be visible)

File: packages/application/src/shell.ts [hunk 1]
```diff
@@ -264,6 +279,22 @@ export class LabShell extends Widget implements JupyterFrontEnd.IShell {
...
+    // Setup single-document-mode title bar
+    const titleWidget = (this._titleWidget = new Widget());
+    titleWidget.id = 'jp-TitlePanel-title';
+    this.add(titleWidget, 'title');
+    if (this._dockPanel.mode == 'multiple-document') {
+      this._titleHandler.panel.hide();
+    }
+
+    this.activeChanged.connect((sender, args) => {
+      let newValue = args.newValue as MainAreaWidget;
```
Comment (afshin) [inline packages/application/src/shell.ts:292]: What happens when the active widget isn't a `MainAreaWidget`? Is this a safe assumption?

File: packages/application/style/titlepanel.css [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#jp-title-panel {
+  min-height: var(--jp-private-title-panel-height);
+  width: 100%;
+  display: flex;
+  background: var(--jp-layout-color1);
+  overflow: visible;
+}
+
+#jp-TitlePanel-title {
```
Comment (afshin) [inline packages/application/style/titlepanel.css:18]: Since this is an ID, should it be `#jjp-title-panel-title`?

Comment (afshin) [reply packages/application/style/titlepanel.css:18]: Hm, it appears we are inconsistent, cf. `#jp-MainLogo`.

File: packages/mainmenu-extension/style/index.css [hunk 3]
```diff
@@ -3,4 +3,20 @@
 | Distributed under the terms of the Modified BSD License.
 |----------------------------------------------------------------------------*/
 
+.jp-LabShell[data-shell-mode='single-document'] #jp-top-panel #jp-MainLogo {
+  animation-duration: 1s;
+  animation-name: logo-slideup;
+  animation-fill-mode: forwards;
+}
+
+@keyframes logo-slideup {
```
Comment (vidartf) [inline packages/mainmenu-extension/style/index.css:12]: I'm not sure if this is needed, as the logo isn't really changing position from the user's point of view (in top left corner of view both before and after). At least that is my impression from the gif, not sure if there are other cases I'm missing?

File: packages/application/src/shell.ts (hunk 1 above)
Comment (ellisonbg) [reply packages/application/src/shell.ts:292]: Fixing with a type guard in an upcoming commit.

File: packages/mainmenu-extension/style/index.css (hunk 3 above)
Comment (ellisonbg) [reply packages/mainmenu-extension/style/index.css:12]: Removing in an upcoming commit

File: packages/application/style/titlepanel.css (hunk 2 above)
Comment (ellisonbg) [reply packages/application/style/titlepanel.css]: Fixed in latest commit.

**Comment by ellisonbg** [comment]
All review comments addressed and additional improvements to document renaming.

**Comment by telamonian** [comment]
Overall, LGTM! My only nit is that the current expand/collapse behavior for the left sidebar is a bit confusing. Current behavior (with this PR):

- on initial page load/reload, regardless of previous state:
  - if in SD mode, left sidebar is collapsed
  - if in normal mode, left sidebar is expanded
- otherwise, the left sidebar ignores document mode and only expands/collapses when a sidebar tab is directly clicked (same as previous behavior)

Ideally, the side bars would instead:
- remember their most recent state wrt to each document mode, and restore said expand/collapse state on document mode switch
- restore to immediate prior state on page reload

If that's too much for this PR, I think a good/simple place to start would be to just have the sidebar collapse on switch into SD mode, and expand on switch into normal mode, just like how the sidebar behaves right now in this PR on mode toggle + page reload.

**Comment by ellisonbg** [comment]
@telamonian I think we should follow up in subsequent PRs to address the left panel layout state. I have a separate PR underway that is diving into how document mode and layout restoration interplay. 

Can we merge this?

**Comment by wolfv** [comment]
Did this get an UX/UI review?
We just hit the problem where jupyterlab dev mode was started in single-view mode and we didn't know how to get out of it / that we were in it. 

It would be great to have an indicator that one is in a special mode and maybe have a button next to the enlarged title to go back to "normal".

Or in the tab manager, one could have a button to switch to multi-doc mode.
Right now it's super-hidden.

**Pull request merged**